### PR TITLE
✨ Gitlab: Add projects to cron

### PR DESCRIPTION
**Pull request by raghavkaul**
#### What kind of change does this PR introduce?

Add cron worker support for GitLab projects. Switch worker's RepoClient when using a GitLab repo.

- [x] Tests for the changes have been added (for bug fixes/features)

### CODE REVIEW

File: cron/k8s/controller.yaml [hunk 1]
```diff
@@ -52,7 +52,11 @@ spec:
           containers:
             - name: controller
               image: gcr.io/openssf/scorecard-batch-controller:stable
-              args: ["--config=/etc/scorecard/config.yaml", "cron/internal/data/projects.csv"]
+              args: [
+                "--config=/etc/scorecard/config.yaml", 
+                "cron/internal/data/projects.csv", 
+                "cron/internal/data/gitlab-projects-selected.csv"
```
Comment (spencerschrock) [inline cron/k8s/controller.yaml:58]: what are your thoughts on doing this in the release controller config instead?

Comment (raghavkaul) [reply cron/k8s/controller.yaml:58]: What is the difference between this controller and the release controller?

Comment (spencerschrock) [reply cron/k8s/controller.yaml:58]: The controller enumerates the weekly 1.2M repos for the production workers.
The release controller is the nightly test that does 5000 repos for the staging worker that moves commits from `latest` to `stable`.

My opinion is that the experimental features should stay in the release/nightly worker to not affect the cron.

File: checker/client.go [hunk 2]
```diff
@@ -58,7 +58,7 @@ func GetClients(ctx context.Context, repoURI, localURI string, logger *log.Logge
 	var repoClient clients.RepoClient
 
 	//nolint:nestif
-	if experimental && glrepo.DetectGitLab(repoURI) {
```
Comment (spencerschrock) [inline checker/client.go:61]: There are reasons to use `SCORECARD_EXPERIMENTAL` for GitHub repos as well. This change forces the gitlab repo client.

File: clients/gitlabrepo/client.go [hunk 3]
```diff
@@ -228,6 +229,22 @@ func (client *Client) Close() error {
 	return nil
 }
 
+func CreateGitlabDotComClient(ctx context.Context) (clients.RepoClient, error) {
+	var err error
+	var gitlabComRepo clients.Repo
+	if gitlabComRepo, err = MakeGitlabRepo("https://gitlab.com/gitlab-org/gitlab"); err != nil {
```
Comment (spencerschrock) [inline clients/gitlabrepo/client.go:235]: Is there significance to this repo? Or is this just an arbitrary repo on the gitlab.com host?

File: clients/gitlabrepo/client.go [hunk 4]
```diff
@@ -285,14 +302,3 @@ func CreateGitlabClientWithToken(ctx context.Context, token string, repo clients
 func CreateOssFuzzRepoClient(ctx context.Context, logger *log.Logger) (clients.RepoClient, error) {
 	return nil, fmt.Errorf("%w, oss fuzz currently only supported for github repos", clients.ErrUnsupportedFeature)
 }
-
-// DetectGitLab: check whether the repoURI is a GitLab URI
-// Makes HTTP request to GitLab API.
-func DetectGitLab(repoURI string) bool {
```
Comment (spencerschrock) [inline clients/gitlabrepo/client.go:291]: regarding the comment in `checker/client.go`, does it make sense to have some sort of gitlab detection still?

File: clients/gitlabrepo/repo.go [hunk 5]
```diff
@@ -46,6 +47,9 @@ type repoURL struct {
 The following input format is not supported:
 	* https://gitlab.<companyDomain:string>.com/projects/<projectID:int>
 */
+
+var errInvalidGitlabRepoURL = errors.New("repo is not a gitlab repo")
+
```
Comment (spencerschrock) [inline clients/gitlabrepo/repo.go:52]: can we move this higher so we don't get between the godoc and function for `parse`?

File: clients/gitlabrepo/repo_test.go [hunk 6]
```diff
@@ -157,9 +157,13 @@ func TestRepoURL_DetectGitlab(t *testing.T) {
 		if tt.flagRequired && os.Getenv("TEST_GITLAB_EXTERNAL") == "" {
 			continue
 		}
-		g := DetectGitLab(tt.repouri)
-		if g != tt.expected {
-			t.Errorf("got %s isgitlab: %t expected %t", tt.repouri, g, tt.expected)
+		g, err := MakeGitlabRepo(tt.repouri)
```
Comment (spencerschrock) [inline clients/gitlabrepo/repo_test.go:160]: If we're switching what the test is testing, lets change the name of the function above as well (wouldn't let me comment up there.)

File: clients/gitlabrepo/repo_test.go [hunk 7]
```diff
@@ -157,9 +157,13 @@ func TestRepoURL_DetectGitlab(t *testing.T) {
...
+		if (g != nil) != (err == nil) {
```
Comment (spencerschrock) [inline clients/gitlabrepo/repo_test.go:161]: should this just be `if err != nil` ? or are we expecting some calls to error?

File: cron/data/iterator.go [hunk 8]
```diff
@@ -83,9 +84,14 @@ func (reader *csvIterator) Next() (RepoFormat, error) {
...
-	if _, err := githubrepo.MakeGithubRepo(reader.next.Repo); err != nil {
-		return reader.next, fmt.Errorf("invalid GitHub URL: %w", err)
+
+	repoURI := reader.next.Repo
+
+	// validate gitlab or github url
+	if _, err := gitlabrepo.MakeGitlabRepo(repoURI); err != nil {
+		if _, err := githubrepo.MakeGithubRepo(repoURI); err != nil {
+			return reader.next, fmt.Errorf("invalid URL, neither github nor gitlab: %w", err)
+		}
```
Comment (spencerschrock) [inline cron/data/iterator.go:94]: this makes a call to the gitlab API correct?

Can we at least switch the ordering around so the cron isn't making 1.2M gitlab requests each week?

Comment (raghavkaul) [reply cron/data/iterator.go:94]: It shouldn't do that; the solution is a little ugly but in `gitlabrepo.isvalid` we exit early if the `repo.host` is `github.com`.

File: cron/internal/data/gitlab-projects-selected.csv [hunk 9]
```diff
@@ -0,0 +1,100 @@
+https://gitlab.com/gitlab-org/gitlab-foss,
```
Comment (spencerschrock) [inline cron/internal/data/gitlab-projects-selected.csv:1]: while we support not having the csv header, lets keep it here for consistency with `projects.csv`

File: cron/internal/data/projects.release.csv [hunk 10]
```diff
@@ -0,0 +1,2501 @@
+repo,metadata
```
Comment (spencerschrock) [inline cron/internal/data/projects.release.csv:1]: this file shouldn't be committed.

File: cron/internal/worker/main.go [hunk 11]
```diff
@@ -131,22 +161,27 @@ func processRequest(ctx context.Context,
 	batchRequest *data.ScorecardBatchRequest,
 	blacklistedChecks []string, bucketURL, rawBucketURL, apiBucketURL string,
 	checkDocs docs.Doc,
-	repoClient clients.RepoClient, ossFuzzRepoClient clients.RepoClient,
+	githubClient, gitlabClient clients.RepoClient, ossFuzzRepoClient clients.RepoClient,
 	ciiClient clients.CIIBestPracticesClient,
 	vulnsClient clients.VulnerabilitiesClient,
 	logger *log.Logger,
 ) error {
+	fmt.Println("reached")
```
Comment (spencerschrock) [inline cron/internal/worker/main.go:169]: this got left in

File: clients/gitlabrepo/repo_test.go (hunk 7 above)
Comment (raghavkaul) [reply clients/gitlabrepo/repo_test.go:161]: There should be an error whenever the result is nil, and there should be no error when the result is non-nil. For every test where expected=false, we expect an error & nil return. It's probably overkill, but this is my pattern to check that the conditions on L164 are either both true or both false.

File: clients/gitlabrepo/client.go (hunk 4 above)
Comment (raghavkaul) [reply clients/gitlabrepo/client.go:291]: I think there's a way to do without detection. `MakeGitlabRepo` and `DetectGitlab` both called `parse` and `isvalid`, so it didn't feel like they needed to be separate functions.

File: cron/worker/worker_test.go [hunk 12]
```diff
@@ -55,3 +55,49 @@ func TestResultFilename(t *testing.T) {
 		})
 	}
 }
+
+func asptr(s string) *string {
+	return &s
+}
+
+func TestGitlabRepo(t *testing.T) {
```
Comment (spencerschrock) [inline cron/worker/worker_test.go:63]: what is this test trying to do?

Comment (spencerschrock) [inline cron/worker/worker_test.go:63]: what is this test trying to do?
it seems like a copy/paste of the test above it with a minimal change that doesnt affect output?

Comment (spencerschrock) [inline cron/worker/worker_test.go:63]: what is this test trying to do?
it seems like a copy/paste of the test above it with a minimal change that doesnt affect output?

File: cron/internal/worker/main.go [hunk 13]
```diff
@@ -143,10 +174,14 @@ func processRequest(ctx context.Context,
 	// TODO: run Scorecard for each repo in a separate thread.
 	for _, repoReq := range batchRequest.GetRepos() {
 		logger.Info(fmt.Sprintf("Running Scorecard for repo: %s", *repoReq.Url))
-		repo, err := githubrepo.MakeGithubRepo(*repoReq.Url)
-		if err != nil {
+		var repo clients.Repo
+		var err error
+		repoClient := githubClient
+		if repo, err = gitlabrepo.MakeGitlabRepo(*repoReq.Url); err != nil {
+			repoClient = gitlabClient
```
Comment (spencerschrock) [inline cron/internal/worker/main.go:181]: same question about this making a gitlab request in the cron, which will result in millions of queries.

File: cron/internal/worker/main.go [hunk 14]
```diff
@@ -161,7 +196,16 @@ func processRequest(ctx context.Context,
 		if err != nil {
 			return fmt.Errorf("error during policy.GetEnabled: %w", err)
 		}
-		for _, check := range blacklistedChecks {
+
+		disabledChecks := blacklistedChecks
+		if _, err := gitlabrepo.MakeGitlabRepo(*repoReq.Url); err != nil {
```
Comment (spencerschrock) [inline cron/internal/worker/main.go:201]: can we save the results from above where we set the gitlab repos? or make this logic be in the same place?

**Comment by spencerschrock** [comment]
Sorry for all the spam, this was supposed to go out as one review.

File: cron/internal/worker/main.go (hunk 13 above)
Comment (raghavkaul) [reply cron/internal/worker/main.go:181]: API queries for client detection shouldn't happen unless the URL is not `github.com|gitlab.com`.

File: clients/gitlabrepo/client.go (hunk 3 above)
Comment (raghavkaul) [reply clients/gitlabrepo/client.go]: It's arbitrary; changed this to use `gitlab.WithBaseURL`

File: cron/worker/worker_test.go (hunk 12 above)
Comment (raghavkaul) [reply cron/worker/worker_test.go]: Yes, it's spurious. Removed

File: checker/client.go (hunk 2 above)
Comment (spencerschrock) [reply checker/client.go:61]: Marking unresolved, as this still looks like it's an issue. Try running this to see my point:
```
SCORECARD_EXPERIMENTAL=1 go run main.go --repo https://github.com/ossf/scorecard
```

File: cron/internal/worker/main.go [hunk 15]
```diff
@@ -91,7 +118,11 @@ func newScorecardWorker() (*ScorecardWorker, error) {
 
 	sw.ctx = context.Background()
 	sw.logger = log.NewLogger(log.InfoLevel)
-	sw.repoClient = githubrepo.CreateGithubRepoClient(sw.ctx, sw.logger)
+	sw.githubClient = githubrepo.CreateGithubRepoClient(sw.ctx, sw.logger)
+	// TODO(raghavkaul): Read GitLab auth token from environment
+	if sw.gitlabClient, err = gitlabrepo.CreateGitlabClient(sw.ctx, "gitlab.com"); err != nil {
```
Comment (spencerschrock) [inline cron/internal/worker/main.go:123]: the baseurls are expected to include the scheme, or you'll get repo unreachable errors for every repo

can we make this a constant with value `https://gitlab.com/`?

File: cron/k8s/controller.release.yaml [hunk 16]
```diff
@@ -52,7 +52,11 @@ spec:
           containers:
             - name: controller
               image: gcr.io/openssf/scorecard-batch-controller:latest
-              args: ["--config=/etc/scorecard/config.yaml", "cron/internal/data/projects.release.csv"]
+              args: [
+                "--config=/etc/scorecard/config.yaml",
+                "cron/internal/data/projects.release.csv",
+                "cron/internal/data/gitlab-projects-selected.csv"
```
Comment (spencerschrock) [inline cron/k8s/controller.release.yaml:58]: I still get panics running this against repos in this list. Lets fix these up before merging in.

**Review by spencerschrock — COMMENTED** [review]
Overall the logic is good, but there are a few blockers to getting this merged. Let's aim for having this run in the cron emulator without errors before merging it in for our nightly release tests.

Also, I personally think we shouldn't check in `gitlab-projects.csv` yet, but not a blocker on my side.

**Review by spencerschrock — APPROVED** [review]
Looks good overall

File: cron/internal/worker/main.go [hunk 17]
```diff
@@ -49,14 +50,40 @@ const (
...
+		"Fuzzing",
+		"License",
+		"Maintained",
+		"Packaging",
+		"Pinned-Dependencies",
+		"SAST",
+		// "Security-Policy",
+		"Signed-Releases",
+		"Token-Permissions",
+		"Vulnerabilities",
```
[the remark this answers is not in the document]
Comment (raghavkaul) [reply cron/internal/worker/main.go:76]: Yes, added

File: cron/k8s/controller.release.yaml [hunk 18]
```diff
@@ -66,6 +70,8 @@ spec:
                   value: "gs://ossf-scorecard-rawdata-releasetest"
                 - name: SCORECARD_SHARD_SIZE
                   value: "5"
+                - name: "SCORECARD_EXPERIMENTAL"
```
[the remark this answers is not in the document]
Comment (raghavkaul) [reply cron/k8s/controller.release.yaml]: Yes, this flag should be set in the worker instead, changed

File: clients/gitlabrepo/contributors_test.go [hunk 19]
```diff
@@ -48,7 +48,7 @@ func Test_ContributorsSetup(t *testing.T) {
 				t.Error("couldn't make gitlab repo", err)
 			}
 
-			client, err := CreateGitlabClientWithToken(context.Background(), "", repo)
+			client, err := CreateGitlabClientWithToken(context.Background(), "", repo.Host())
```
[the remark this answers is not in the document]
Comment (raghavkaul) [reply clients/gitlabrepo/contributors_test.go:51]: Will create a follow up PR to move this testing to `e2e/`.

**Pull request merged**